### PR TITLE
add problems-widget filter toolbar

### LINKED ISSUE #5135: [markers] add problems-widget filter toolbar item
**Description**

Currently, there is no way to filter `problem markers` from the `problems-widget`.
I propose we implement filtering of markers by **error type** or by **text matching** similarly to vscode.

<div align='center'>

Screen Shot 2019-05-13 at 3 37 00 PM

</div>

### PR DESCRIPTION
**Pull request by gaomin**
#### What it does

Closed #5135.

I refer to vscode and implement the problems-widget filter feature. The function description is as follows：
1.  Supports filtering in these ways：
Filter input text，"Show Errors", "Show Warnings", "Show Infos", "Show Hints", "Show Active File Only", "Hide Excluded Files".
Filter input text, filter by message match or file path match.
"Show Errors", "Show Warnings", "Show Infos", "Show Hints", filter by severity.
"Show Active File Only", filter by current editor file.
"Hide Excluded Files", reverse match of input box text result.

2. Filter history 
Filter history apply when press up or down key.

截屏2023-03-27 下午2 25 25

#### How to test

1. Start Theia 
2. Open a project
3. Open Problems
4. Enter some text in filter input or change filter option

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/markers/src/browser/marker-manager.ts [hunk 1]
```diff
@@ -202,4 +221,61 @@ export abstract class MarkerManager<D extends object> {
...
+            let filterText = text.replace(/^!/, '').replace(/\/$/, '');
+            if (filterText[0] === '.') {
+                filterText = '*' + filterText; // convert ".js" to "*.js"
+            }
+
+            if (startWithExclude !== useFilesExclude) {
+                markfilterFns.push((data: D, collection: MarkerCollection<D>) => !((data as Diagnostic).message.toLowerCase().includes(text.replace(/^!/, '').toLowerCase()) ||
+                    [`**/${filterText}`, `**/${filterText}/**`].some((t: string) => match(t, collection.uri.path.toString()))));
+            } else {
+
```
Comment (vince-fugnitto) [inline packages/markers/src/browser/marker-manager.ts:269]: minor: unnecessary newline.

File: packages/markers/src/browser/marker-manager.ts [hunk 2]
```diff
@@ -110,22 +113,31 @@ export interface Owner2MarkerEntry {
 
 @injectable()
 export abstract class MarkerManager<D extends object> {
-
     public abstract getKind(): string;
-
+    protected toolbarFilters: Filters;
+    protected toolbarFilterFn: (data: D, collection: MarkerCollection<D>) => boolean = () => true;
```
Comment (vince-fugnitto) [inline packages/markers/src/browser/marker-manager.ts:118]: minor: we should use a better name.

**Review by vince-fugnitto — COMMENTED** [review]
Thank you for the contribution! :+1: 

I have some comments overall on the code but I'll perform a more in depth-review later on when they are addressed and cleaned up.

I also noticed some styling differences when quickly testing which should be addressed to more closely resemble the behavior in vscode:
- the input outline is incorrect
- the hover color for the filter is incorrect
- the styling for "showing {0} of {1}" message is incorrect

**Comment by gaomin** [comment]
hello, I fixed problems and please review it, thanks

**Comment by tsmaeder** [comment]
Ping @vince-fugnitto?

File: packages/markers/src/browser/problem/problem-filter.tsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            { key: 'useFilesExclude', label: 'Hide Excluded Files', group: '1' }
+        ];
+
+        return opts.map(({ key, label, group }) => ({
+            value: key,
+            label: creatLabel(key === 'showHints' ? nls.localize('theia/markers/showHints', label) : nls.localizeByDefault(label), filters[key as FilterKeys]),
+            group,
+            onClick: () => clickHandler(key as FilterKeys)
+        }));
+    }, [filters]);
```
Comment (msujew) [inline packages/markers/src/browser/problem/problem-filter.tsx:122]: Suggestion: This makes it a bit hard for the static analysis to confirm the existance of the localization keys. I would suggest the following:

```suggestion
    const options = React.useMemo(() => {
        const opts = [
            { key: 'showErrors', label: nls.localizeByDefault('Show Errors'), group: '0' },
            { key: 'showWarnings', label: nls.localizeByDefault('Show Warnings'), group: '0' },
            { key: 'showInfos', label: nls.localizeByDefault('Show Infos'), group: '0' },
            { key: 'showHints', label: nls.localize('theia/markers/showHints', 'Show Hints'), group: '0' },
            { key: 'activeFile', label: nls.localizeByDefault('Show Active File Only'), group: '1' },
            { key: 'useFilesExclude', label: nls.localizeByDefault('Hide Excluded Files'), group: '1' }
        ];

        return opts.map(({ key, label, group }) => ({
            value: key,
            label: creatLabel(label, filters[key as FilterKeys]),
            group,
            onClick: () => clickHandler(key as FilterKeys)
        }));
    }, [filters]);
```

File: packages/core/src/browser/widgets/dropdown-component.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+export interface DropdownComponentProps {
+    options: DropdownOption[];
+    children: React.ReactElement;
+    onBlur?: () => void,
+    onFocus?: () => void
+}
+
+export const DROPDOWN_COMPONENT_CONTAINER = 'dropdown-component-container';
+
+export const DropdownComponent: React.FC<DropdownComponentProps> = props => {
```
Comment (msujew) [inline packages/core/src/browser/widgets/dropdown-component.tsx:37]: This dropdown doesn't align with the CSS of our other menus. Would there be a way to automatically apply the styling of the phosphor menus to this dropdown or reuse those html elements instead?

**Comment by msujew** [comment]
@gaomin Do you plan on getting back to this?

**Comment by msujew** [comment]
I'll close this. Please ping us in case you plan to get back to this PR @gaomin. Anyone else is free to pick up this work and create a new PR.

**Pull request closed without merging**